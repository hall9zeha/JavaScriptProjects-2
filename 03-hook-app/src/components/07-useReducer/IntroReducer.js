//Creamos un lista de tareas con un objeto
import React from 'react'

export const IntroReducer = () => {

const initialObjectState = [{
    id:123,
    todo:'any task',
    done:false
}]

//Creamos nuestra función reducer que manejará los casos de uso
//Por convención una función Reducer debe tener dos argumentos, un estado y una acción
const todoReducer = (state = initialObjectState, action)=>{
    //debe retornar un nuevo valor u objeto y no modificar el valor u objeto del estado inicial, este debe ser inmutable
    if(action?.type ==='add'){//evalueamos que tipo de acción es y si esta nula o no
        return [...state,action.payLoad]//retornamos una copia del estado anterior y el nuevo objeto agregado
    }
    return state

}

//Creamos un nuevo objeto que será nuestra segunda tarea para agregar a una supuesta lista de tareas
const newTodoObject = {
    id:456,
    todo:'other task',
    done:false
}

//obtenemos el último valor devuelto por nuestro Reducer
let todos = todoReducer();

//Ahora  la acción que recibirá nuestra función Reducer será un objeto
const todoActionAdd={
    type:'add',//debe contener un tipo de acción
    payLoad:newTodoObject//Y el objeto o valor
}

//Enviamos los argumentos a nuestro Reducer
const todo = todoReducer(todos, todoActionAdd);
console.log(todo);


  return (
    <div>
        <h1>IntroReducer</h1>
        <hr/>
        <h3>Abre la consola para ver el resultado</h3>

    </div>
    
  )
}
