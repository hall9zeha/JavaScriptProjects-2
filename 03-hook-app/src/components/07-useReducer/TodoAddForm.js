import React, { Fragment } from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAddForm = ({handleTodoAdd}) => {


    const [{description}, handleInputChange,resetForm]=useForm({
        description:''
    })


    const handleSubmit = (e)=>{
        e.preventDefault();
        //Si el input box está vacío no hacer nada
        if(description.trim().length <=1){
            return;
        }
        const newTodo = {
            id:new Date().getTime(),
            desc:description,
            done:false
        };

        handleTodoAdd(newTodo);
        resetForm();
    }

  return (
    <>
         <h4>Agregar TODO</h4>
                <hr/>
                <form onSubmit={handleSubmit}>
                <input
                    type='text'
                    name='description'
                    className='form-control'
                    placeholder='Escribe una tarea'
                    autoComplete='off'
                    value={description}
                    onChange={handleInputChange}
                />
                <button type='submit' className='btn btn-block btn-outline-primary  mt-1 '>Agregar</button>
                </form>
    </>
  )
}
