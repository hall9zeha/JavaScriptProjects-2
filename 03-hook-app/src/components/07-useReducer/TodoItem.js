import React from 'react'

export const TodoItem = ({todo,index, handleDoneTask, handleDelete}) => {
  return (
    <li key={todo.id}
        className='list-group-item'
        >
        <p className={`${todo.done && 'complete'}`}
            onClick={()=>{handleDoneTask(todo.id)}}
        >{index+1 }.{todo.desc}</p>
        <button 
            data-testid='btn-delete'
            onClick={()=>{handleDelete(todo.id)}}
            className='btn btn-danger'>Borrar</button> 
    </li>
  )
}
