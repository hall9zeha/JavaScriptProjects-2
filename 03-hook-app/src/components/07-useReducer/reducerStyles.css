body{
    padding: 50px;
}
p{
    margin: 0px;
}
.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
}
.complete{
    text-decoration:line-through;
}
.btn-block{
    display: block;
    width: 100%;
}