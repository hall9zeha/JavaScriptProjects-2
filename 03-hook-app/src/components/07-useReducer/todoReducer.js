

export const todoReducer = (state = [], action ) => {
  switch (action.type) {
    case 'add':
        return [...state,action.payload];//ya que retornamos directamente no necesitamos  break
        
    default:
        return state;
  }
}
