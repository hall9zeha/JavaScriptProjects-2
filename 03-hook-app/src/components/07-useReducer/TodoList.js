import React from 'react'
import { TodoItem } from './TodoItem'

export const TodoList = ({todos, handleDoneTask, handleDelete}) => {
  return (
    
             <ul className='list-group list-group-flush'>
                {
                    todos.map((todo, i)=>(
                      <TodoItem
                        key={todo.id}
                        todo={todo} 
                        index={i} 
                        handleDoneTask={handleDoneTask} 
                        handleDelete={handleDelete}/>
                    ))
                }
                </ul>
   
  )
}
