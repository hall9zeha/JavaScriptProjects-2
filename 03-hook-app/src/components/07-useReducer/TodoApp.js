import React, { useEffect, useReducer } from 'react'

import './reducerStyles.css'
import { TodoAddForm } from './TodoAddForm'
import { TodoList } from './TodoList'
import { todoReducer } from './todoReducer'

const init=()=>{
    // return [{
    //     id:new Date().getTime(),
    //     desc:'learning Android multiplatform',
    //     done:false
    // }]
    //Usaremos lo que almacenemos en Local storage
    return JSON.parse(localStorage.getItem('todos')) || []; //si existen retornar, sino un arreglo vacío
}
export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [],init)
    


    //useEffect para lanzar la orden de grabar en Local storage
    useEffect(()=>{

        localStorage.setItem('todos', JSON.stringify(todos))
    },[todos]);//cuando cambie la lista de tareas se guardará una nueva versión en LocalStorage

    const handleTodoAdd = (newTodo)=>{
        dispatch({
            type:'add',
            payload:newTodo
            }
        )
    }

    //Para eliminar un registro
    const handleDelete = (todoId)=>{
        console.log(todoId);
        const actionDelete = {
            type:'delete',
            payload:todoId
        }
        dispatch(actionDelete);

    }
    const handleDoneTask =(todoId)=>{
        const actionDone={
            type:'done',
            payload:todoId
        }
        dispatch(actionDone);
    }

  return (
    <div>
        <h1>TodoApp ({todos.length})</h1>
        <hr/>
        <div className='row'>
            <div className='col-7'>
                <TodoList 
                    todos={todos}
                    handleDoneTask={handleDoneTask}
                    handleDelete={handleDelete}/>
                
            </div>
            <div className='col-5'>
               <TodoAddForm handleTodoAdd={handleTodoAdd} />
            </div>
        </div>
        
    </div>
  )
}
