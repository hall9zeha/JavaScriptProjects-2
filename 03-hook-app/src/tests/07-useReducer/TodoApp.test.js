
import { prettyDOM, render, screen} from '@testing-library/react';
import React from 'react';

import renderer, { act } from "react-test-renderer";
import { TodoAddForm } from '../../components/07-useReducer/TodoAddForm';
import { TodoApp } from '../../components/07-useReducer/TodoApp';
import { TodoList } from '../../components/07-useReducer/TodoList';
import { todos } from '../../fixtures/DemoTodos';


describe('Tests in <TodoApp/> component', () => { 
    let wrapper = renderer.create(<TodoApp/>);
    Storage.prototype.setItem = jest.fn();


    test('should show successfully',()=>{

        expect(wrapper.toJSON()).toMatchSnapshot();
    })

    test('should add a TODO',()=>{
     
        
        const root = wrapper.root;
        //obtenemos una instancia de la función dentro del componente
        const handle=root.findByType(TodoAddForm).props.handleTodoAdd
        
        act(()=>{
            //Enviamos dos TODOS
            handle(todos[0])
            handle(todos[1])

        });
        
        const h1 = root.findByType('h1').props.children.join('');
             
        console.log(handle)
        //La etiqueta h1 debe tener el siguiente texto
        expect(h1).toBe('TodoApp (2)')
        //Debe de realizar dos inserciones en el LocalStorage
        expect(localStorage.setItem).toHaveBeenCalledTimes(2)
     
    })
    test('should delete a TODO',()=>{
        const wrapper = renderer.create(<TodoApp/>);
        const root = wrapper.root;
        //obtenemos una instancia de la función dentro del componente
        const handle=root.findByType(TodoAddForm).props.handleTodoAdd
        const handleDel=root.findByType(TodoList).props.handleDelete
        
        act(()=>{
      
            handle(todos[0])
            handleDel(todos[0].id)
        });
     
        const h1 = root.findByType('h1').props.children.join('');
        expect(h1).toBe('TodoApp (0)')
    })
 })