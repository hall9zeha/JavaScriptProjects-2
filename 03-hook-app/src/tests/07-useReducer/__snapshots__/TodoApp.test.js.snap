// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <TodoApp/> component should show successfully 1`] = `
<div>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <ul
        className="list-group list-group-flush"
      />
    </div>
    <div
      className="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="description"
          onChange={[Function]}
          placeholder="Escribe una tarea"
          type="text"
          value=""
        />
        <button
          className="btn btn-block btn-outline-primary  mt-1 "
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
